lib/chat_app_web/live/single_room_live.ex

```diff
@@ -45,14 +48,30 @@ defmodule ChatAppWeb.SingleRoomLive do
 
   def handle_event("delete_message", %{"id" => message_id_from_client}, socket) do
     IO.puts message_id_from_client
-    broadcasted_message = {:delete_messages, message_id_from_client}
-    PubSub.broadcast(ChatApp.PubSub, socket.assigns.topic, broadcasted_message)
+    {:noreply, assign(socket, messages: Enum.map(socket.assigns.messages, fn {message_id, message} ->
+      cond do
+        message_id == message_id_from_client -> {message_id, %{message | is_deleted: true}}
+        true -> {message_id, message}
+      end
+    end))}
+  end
+
+  def handle_event("user_typing_indication", _, socket) do
+    Logger.info("#{socket.assigns.username} is typing...")
+    PubSub.broadcast(ChatApp.PubSub, "users_typing", {:users_typing, socket.assigns.username, true})
+    {:noreply, socket}
+  end
+
+  def handle_event("user_typing_indication_ended", _, socket) do
+    Logger.info("#{socket.assigns.username} is not typing anymore.")
+    PubSub.broadcast(ChatApp.PubSub, "users_typing", {:users_typing, socket.assigns.username, false})
     {:noreply, socket}
   end
```

Comment:
Why not just keep a list of the users who are typing, instead of storing them all and setting a boolean flag on them that reflects whether they are typing or not?

Reply:
I assumed it would take O(n) time complexity to search within a list, while searching a user in a map should be a O(1) operation. I'm definitely not sure that is correct, but I will change it into a list.